Errors when creating problems where some variables do not have equations

Two MWEs, one where value is provided through initial condition and one through default value.
```julia
using ModelingToolkitBase
using ModelingToolkitBase: t_nounits as t, D_nounits as D
ps = @parameters p d
us = @variables X(t) Y(t) = 1.0
eq = [
    D(X) ~ p - d*X
]
@mtkbuild sys = System(eq, t, us, ps)
ODEProblem(sys, [X => 0.0, p => 2.0, d => 0.1], (0.0, 1.0))

using ModelingToolkitBase
using ModelingToolkitBase: t_nounits as t, D_nounits as D
ps = @parameters p d
us = @variables X(t) Y(t)
eq = [
    D(X) ~ p - d*X
]
@mtkbuild sys = System(eq, t, us, ps)
ODEProblem(sys, [X => 0.0, Y => 1.0, p => 2.0, d => 0.1], (0.0, 1.0))
```
both yields 
```
ERROR: AssertionError: ik in init_ps
Stacktrace:
  [1] timevaring_initsys_process_op!(init_vars_set::ModelingToolkitBase.AtomicArraySet{…}, init_ps::ModelingToolkitBase.AtomicArraySet{…}, eqs_ics::Vector{…}, op::ModelingToolkitBase.AtomicArrayDict{…}, derivative_rules::ModelingToolkitBase.DerivativeDict{…}, guesses::ModelingToolkitBase.AtomicArrayDict{…})
    @ ModelingToolkitBase C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\nonlinear\initializesystem.jl:398
  [2] generate_initializesystem_timevarying(sys::System; op::ModelingToolkitBase.AtomicArrayDict{…}, initialization_eqs::Vector{…}, guesses::Dict{…}, default_dd_guess::Bool, fast_path::Bool, algebraic_only::Bool, check_units::Bool, check_defguess::Bool, name::Symbol, kwargs::@Kwargs{})
    @ ModelingToolkitBase C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\nonlinear\initializesystem.jl:159
  [3] generate_initializesystem_timevarying
    @ C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\nonlinear\initializesystem.jl:46 [inlined]
  [4] #generate_initializesystem#1044
    @ C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\nonlinear\initializesystem.jl:35 [inlined]
  [5] generate_initializesystem
    @ C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\nonlinear\initializesystem.jl:32 [inlined]
  [6] InitializationProblem{…}(sys::System, t::Float64, op::ModelingToolkitBase.AtomicArrayDict{…}; fast_path::Bool, guesses::Dict{…}, check_length::Bool, warn_initialize_determined::Bool, initialization_eqs::Vector{…}, fully_determined::Nothing, check_units::Bool, allow_incomplete::Bool, algebraic_only::Bool, time_dependent_init::Bool, initsys_mtkcompile_kwargs::@NamedTuple{}, kwargs::@Kwargs{…})
    @ ModelingToolkitBase C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\problems\initializationproblem.jl:54
  [7] InitializationProblem
    @ C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\problems\initializationproblem.jl:20 [inlined]
  [8] #_#935
    @ .\none:-1 [inlined]
  [9] maybe_build_initialization_problem(sys::System, iip::Bool, op::ModelingToolkitBase.AtomicArrayDict{…}, t::Float64, guesses::Dict{…}; time_dependent_init::Bool, u0_constructor::Function, p_constructor::Function, floatT::Type, initialization_eqs::Vector{…}, use_scc::Bool, eval_expression::Bool, eval_module::Module, implicit_dae::Bool, kwargs::@Kwargs{…})
    @ ModelingToolkitBase C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\problem_utils.jl:1059
 [10] process_SciMLProblem(constructor::Type, sys::System, op::Vector{…}; build_initializeprob::Bool, implicit_dae::Bool, t::Float64, guesses::Dict{…}, warn_initialize_determined::Bool, initialization_eqs::Vector{…}, eval_expression::Bool, eval_module::Module, fully_determined::Nothing, check_initialization_units::Bool, u0_eltype::Nothing, tofloat::Bool, u0_constructor::typeof(identity), p_constructor::typeof(identity), check_length::Bool, symbolic_u0::Bool, warn_cyclic_dependency::Bool, circular_dependency_max_cycle_length::Int64, circular_dependency_max_cycles::Int64, initsys_mtkcompile_kwargs::@NamedTuple{}, substitution_limit::Int64, use_scc::Bool, time_dependent_init::Bool, algebraic_only::Bool, missing_guess_value::ModelingToolkitBase.MissingGuessValue.var"typeof(MissingGuessValue)", allow_incomplete::Bool, is_initializeprob::Bool, kwargs::@Kwargs{…})
    @ ModelingToolkitBase C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\systems\problem_utils.jl:1391
 [11] (ODEProblem{…})(sys::System, op::Vector{…}, tspan::Tuple{…}; callback::Nothing, check_length::Bool, eval_expression::Bool, expression::Type, eval_module::Module, check_compatibility::Bool, kwargs::@Kwargs{})
    @ ModelingToolkitBase C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\problems\odeproblem.jl:81
 [12] ODEProblem
    @ C:\Users\Torkel\.julia\packages\ModelingToolkitBase\fjBdh\src\problems\odeproblem.jl:73 [inlined]
 [13] ODEProblem
    @ .\none:-1 [inlined]
 [14] ODEProblem(sys::System, op::Vector{Pair{Num, Float64}}, tspan::Tuple{Float64, Float64})
    @ ModelingToolkitBase .\none:-1
```

--- a/lib/ModelingToolkitBase/src/systems/systems.jl
+++ b/lib/ModelingToolkitBase/src/systems/systems.jl
@@ -147,6 +147,16 @@ function __mtkcompile(sys::AbstractSystem;
     filter!(all_dvs) do v
         v in original_vars || split_indexed_var(v)[1] in original_vars
     end
+
+    new_binds = copy(parent(bindings(sys)))
+    new_ics = copy(initial_conditions(sys))
+    for unused in setdiff(original_vars, all_dvs)
+        arr, isarr = split_indexed_var(unused)
+        arr in all_dvs && continue
+        delete!(new_binds, arr)
+        delete!(new_ics, arr)
+    end
+
     setdiff!(all_dvs, inputs, disturbance_inputs)
     if fully_determined === nothing
         fully_determined = false
@@ -342,6 +352,8 @@ function __mtkcompile(sys::AbstractSystem;
     @set! sys.outputs = outputs
     @set! sys.schedule = schedule
     @set! sys.isscheduled = true
+    @set! sys.bindings = new_binds
+    @set! sys.initial_conditions = new_ics
     return sys
 end
 

--- a/lib/ModelingToolkitBase/test/initial_values.jl
+++ b/lib/ModelingToolkitBase/test/initial_values.jl
@@ -344,3 +344,23 @@ end
     sol = solve(prob2)
     @test SciMLBase.successful_retcode(sol)
 end
+
+@testset "metadata defaults for variables removed by `mtkcompile` are removed" begin
+    @parameters p d
+    @variables X(t) Y(t) = 1.0
+    eq = [
+        D(X) ~ p - d*X
+    ]
+    @named sys = System(eq, t, [X, Y], [p, d])
+    @test haskey(initial_conditions(sys), Y)
+    sys = mtkcompile(sys)
+    @test !haskey(initial_conditions(sys), Y)
+    @test_nowarn ODEProblem(sys, [X => 0.0, p => 2.0, d => 0.1], (0.0, 1.0))
+    
+    @variables X(t) Y(t) = X
+    @named sys = System(eq, t, [X, Y], [p, d])
+    @test haskey(bindings(sys), Y)
+    sys = mtkcompile(sys)
+    @test !haskey(bindings(sys), Y)
+    @test_nowarn ODEProblem(sys, [X => 0.0, p => 2.0, d => 0.1], (0.0, 1.0))
+end

